Add more detailed instructions to the dev docs

hmstepanek opened the pull request:
* Include prerequisites.
* Add an example to Vizier section of running unit tests.
* Link to instructions for spinning up a Minikube cluster to deploy onto.
* Clarify the differences between Vizier and Pixie Cloud.
* Add workaround instructions for failed px deploy.
* Note when and where various commands in the instructions should be run and explain
what they do in greater detail.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pixie-io/pixie?pullRequest=324) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pixie-io/pixie?pullRequest=324) before we can accept your contribution.<br/><hr/>**Hannah Stepanek** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pixie-io/pixie?pullRequest=324) it.</sub>

nserrino commented on DEVELOPMENT.md:
```diff
@@ -9,21 +9,47 @@ The Pixie project currently consists of three main components:
 This document outlines the process for setting up the development environment for each of these components.
 
 ## Setting up the Environment
+There are two different ways to run a development environment. The first and easiest
```
Can you add `--network host` to the run_docker.sh file ? This way, minikube should work and we won't have to make any particular recommendations of which Kubernetes platform uses which development environment here. 

docker run --rm -it \
  "${configs[@]}" \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -v "${workspace_root}:/pl/src/px.dev/pixie" \
  --network host \
  "${PX_RUN_DOCKER_EXTRA_ARGS[@]}" \
  "${docker_image_with_tag}" \
  "${exec_cmd[@]}"

nserrino commented on DEVELOPMENT.md:
```diff
@@ -9,21 +9,47 @@ The Pixie project currently consists of three main components:
+To setup the developer environment required to start building Pixie's components,
+it is easiest to run the `run_docker.sh` script. The following script will run the
+Docker container and dump you out inside the docker container console from which
+you can run all the necessary tools to build, test, and deploy Pixie.
```
might want to say deploy Pixie in development mode here, just so readers don't think these steps are necessary to deploy "normal" Pixie

nserrino commented on DEVELOPMENT.md:
```diff
@@ -9,21 +9,47 @@ The Pixie project currently consists of three main components:
+If you plan on using a Minikube environment, launch the dev environment with Minikube,
+via Make.
+
+1. Since this runs via Minikube, you must have Minikube installed. To install it follow
```
Can we note/link the docs.px.dev minikube setup instructions here?  The reason is, the default minikube setup will not have enough memory. https://docs.px.dev/installing-pixie/setting-up-k8s/minikube-setup/

nserrino commented on DEVELOPMENT.md:
````diff
@@ -9,21 +9,47 @@ The Pixie project currently consists of three main components:
+1. Since this runs via Minikube, you must have Minikube installed. To install it follow
+the instructions [here](https://minikube.sigs.k8s.io/docs/start/).
+
+1. Run make to spin up a Minikube environment:
+    ```
+    make dev-env-start
+    ```
````
I feel like we should just delete this entire way of development -- seeing as how stale it is lol. WDYT @aimichelle ? Also @hmstepanek  It's ok if you don't want to delete it in this PR  -- but we should probably not mention it.

hmstepanek replied:
I'm absolutely fine with deleting it here if that's what you would like to do. :)

nserrino replied:
sounds great!

nserrino commented on DEVELOPMENT.md:
```diff
@@ -9,21 +9,47 @@ The Pixie project currently consists of three main components:
 
 
 ## Pixie Cloud
+Pixie Cloud manages users, authentication, and proxying “passthrough” mode. If you want
+to make changes to it then you will need to spin up a self-hosted version to test those
+changes rather than pointing to the official cloud hosted version.
```
Let's say, "If you want to make changes to Pixie Cloud, then you will need to spin up a self-hosted version in development mode to test those changes. If you aren't changing Pixie Cloud, feel free to use officially released Pixie Cloud options listed in our [Install Guides](https://docs.px.dev/installing-pixie/install-guides/)."

nserrino commented on DEVELOPMENT.md:
```diff
@@ -93,25 +119,66 @@ After which, you can rerun a `px auth login` to authenticate with the dev cloud
+### Getting Started
+Verify that you can build Pixie Vizier and run the unit tests.
+
+The following will build Pixie Vizier and then run the unit tests for the UDF
+(User Defined Function) math operations. These UDFs will be available directly off
+the px object inside a PxL script. If you are running the development environment via
+Docker, the following should be run inside the Docker container.
```
I think this may be more information than is needed for people not working on UDFs, what about, 

"The following will run an example unit test in Pixie's Vizier module."

nserrino commented on DEVELOPMENT.md:
```diff
@@ -93,25 +119,66 @@ After which, you can rerun a `px auth login` to authenticate with the dev cloud
+version of Vizier.
 
-Deploying a development version of Vizier is a 2-step process. A release-version of Vizier must first be deployed (through the CLI or YAMLs), which will set up specific cluster-secrets that are not deployed via Skaffold. Skaffold can then be run to build and deploy a version of Vizier based on local changes.
+1. Since both of these steps require a Kubernetes cluster to deploy to, you must have
+Minikube installed and running. Follow the instructions
+[here](https://docs.px.dev/installing-pixie/setting-up-k8s/minikube-setup/). Note that
+both deployment scripts below use the output of `kubectl config current-context` to
+determine which Kubernetes cluster to deploy to. So make sure if you have multiple
+clusters, the context is pointing to the correct target cluster.
```
Technically running a development version of Pixie CLoud requires a Kubernetes cluster as well, so this also applies to the Pixie Cloud section -- can we move it up?

nserrino commented on DEVELOPMENT.md:
````diff
@@ -93,25 +119,66 @@ After which, you can rerun a `px auth login` to authenticate with the dev cloud
     ```
 
-2. Deploy Vizier through CLI/YAML. If deploying Vizier to a dev cloud instance, make sure to export the necessary dev-specific environment variables mentioned in `Pixie Cloud` above.
+1. Install the Pixie CLI and signup for Pixie Community Cloud. Follow steps 1 and 3 on
````
We can't actually tell people to sign up for Community Cloud for Pixie since that would be exclusionary to self-hosted Pixie users. I would tell them to just follow one of the entire guides here: https://docs.px.dev/installing-pixie/install-guides/ (which also includes the `px deploy` step so you can remove it below) and ensure that they can successfullly a script in Pixie either via the UI or CLI.

hmstepanek replied:
I wasn't sure where to put the extra info here about what `px deploy` does and what to do if it fails so I've combined it into this step here. A more appropriate location might be to actually move this into the docs we link to. If you agree, I'll remove it from this PR and have Elaine add it into her pixie-doc PR she's working on.

nserrino reviewed (changes requested):
Hi there -- thank you so much for your submission! It really improves the clarity and context in this file, so we appreciate your help. 

One more note beyond what's in the comments: I noticed that we don't put any extra line breaking in the previous version, so we may want to just keep with that style (and remove the line breaks here) in order to be consistent. However, I am persuadable on this point if there is a particular reason to add the extra line breaks.

zasgar commented on DEVELOPMENT.md:
````diff
@@ -93,25 +119,66 @@ After which, you can rerun a `px auth login` to authenticate with the dev cloud
+Docker, the following should be run inside the Docker container.
+
+```
+bazel test //src/carnot/funcs/builtins:math_ops_test --test_output=errors -j 2
+```
````
Things might more around so I'm not sure we want to list a specific test. We can also just run:

`bazel test //...` 

to make sure all unit tests pass. 

You can use -j $(nproc) which will allow this to scale to the number of CPUs you have available.

hmstepanek commented:
@nserrino Typically I'm used to adding line breaks in order to make the file more readable when editing. Most languages have this as part of linting rules and I generally apply this same principle to markdown docs as well though I'm happy to conform to the existing format. For now I'll just conform to the existing style and if we decide to do something different we can change it later and apply it to the entire file(s).

hmstepanek commented:
So I'm just waiting on a couple points of feedback after making the fixups based on review feedback:
1. Whether we have agreement from @aimichelle to go ahead and delete the alternative make dev env setup instructions.
1. Whether we would like to move the additional notes about `px deploy` out of here and into the pixie docs or not @nserrino ?

And thank you all for your quick review feedback!!

nserrino commented on DEVELOPMENT.md:
```diff
@@ -93,25 +110,34 @@ After which, you can rerun a `px auth login` to authenticate with the dev cloud
 
-2. Deploy Vizier through CLI/YAML. If deploying Vizier to a dev cloud instance, make sure to export the necessary dev-specific environment variables mentioned in `Pixie Cloud` above.
-
+1. Install the Pixie CLI and run `px deploy`. Depending on whether you are pointing to a self-hosted Pixie Cloud or the official Comminity Pixie Cloud follow the appropriate installation guide [here](https://docs.px.dev/installing-pixie/install-guides/). `px deploy` will set up specific cluster-secrets, etc that are not deployed via Skaffold. Wait for this command to successfully complete and Viz ...
```
typo: Comminity->Community

nserrino commented on DEVELOPMENT.md:
```diff
@@ -93,25 +110,34 @@ After which, you can rerun a `px auth login` to authenticate with the dev cloud
+Deploying a development version of Vizier is a 2-step process. An official release-version of Vizier must first be deployed (through the Pixie CLI or YAMLs) and Skaffold can then be run to build and deploy a local development version of Vizier.
 
-1. If deploying Vizier pointing to a dev cloud instance, make sure to export the following environment variables:
+1. If you wish to test development changes made to Pixie Cloud and Vizier, export the following environment variables that will point to the development Pixie Cloud instance:
```
might want to add "both" here just so someone doesn't take accidentally take it as a logical OR

nserrino commented:
> So I'm just waiting on a couple points of feedback after making the fixups based on review feedback:
> 
> 1. Whether we have agreement from @aimichelle to go ahead and delete the alternative make dev env setup instructions.
> 2. Whether we would like to move the additional notes about `px deploy` out of here and into the pixie docs or not @nserrino ?
> 
> And thank you all for your quick review feedback!!

Will let Michelle take #1 

As for #2 - I think the need to manually delete px-operator should be fixed in the latest CLI release, which should release today. So we may be able to omit it and not have the question anymore.

Looks like all the numbers are still 1. instead of 1.,2., etc but otherwise looks to be in good shape

htroisi commented:
On point #2 - I can confirm that the `px delete` command in CLI version 0.7.6 will now delete the `px-operator` pods.

hmstepanek commented:
I removed the instructions about `px delete`, fixed the spelling error (good catch!), and converted to using ordered numbers instead of the generic `1.`.

I didn't realize this till just now but, according to [this](https://www.markdownguide.org/basic-syntax/#ordered-lists) even out of order numbers render as in order numbers. 🤩 
```
1. First item
8. Second item
3. Third item
5. Fourth item
```

1. First item
8. Second item
3. Third item
5. Fourth item

nserrino commented on DEVELOPMENT.md:
```diff
@@ -9,21 +9,38 @@ The Pixie project currently consists of three main components:
 This document outlines the process for setting up the development environment for each of these components.
 
 ## Setting up the Environment
+There are two different ways to run a development environment. The first and easiest approach is via Docker. If you plan to use a non-Minikube environment (GKE, EKS, etc) for running Vizier/Pixie Cloud, use this first approach via Docker. The second approach is using Make and Minikube.
```
This will change because we are removing references to the make way, right? Also, I don't think there is a GKE/EKS/Minikube association either way with either approach

hmstepanek replied:
I've deleted the second method. This is all inside a single commit so if we need to role it back we can.

nserrino commented on DEVELOPMENT.md:
```diff
@@ -93,25 +110,30 @@ After which, you can rerun a `px auth login` to authenticate with the dev cloud
+2. Install the Pixie CLI and run `px deploy`. Depending on whether you are pointing to a self-hosted Pixie Cloud or the official Community Pixie Cloud follow the appropriate installation guide [here](https://docs.px.dev/installing-pixie/install-guides/). `px deploy` will set up specific cluster-secrets, etc that are not deployed via Skaffold. Wait for this command to successfully complete and Viz ...
 
-4. Deploy a local version of Vizier using skaffold.
+3. Deploy a local development version of Pixie Vizier using Skaffold. Note each time you make a code change you will need to run this command to build and deploy the new version.
```
Minor but we might want to link to skaffold's docs like [Skaffold](https://skaffold.dev/) and same with bazel
